homeassistant/auth/providers/openid.py

```diff
@@ -0,0 +1,282 @@
+"""OpenID based authentication provider."""
+
+import logging
+import re
+from secrets import token_hex
+from typing import Any, Dict, Optional, cast
+
+from aiohttp import ClientResponseError
+from aiohttp.client import ClientResponse
+import voluptuous as vol
+from yarl import URL
+
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.config_entry_oauth2_flow import (
+    AUTH_CALLBACK_PATH,
+    _encode_jwt,
+    async_register_view,
+)
+from homeassistant.helpers.network import get_url
+
+from . import AUTH_PROVIDER_SCHEMA, AUTH_PROVIDERS, AuthProvider, LoginFlow
+from ..models import Credentials, UserMeta
+
+REQUIREMENTS = ["python-jose==3.1.0"]
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_CLIENT_ID = "client_id"
+CONF_CLIENT_SECRET = "client_secret"
+CONF_CONFIGURATION = "configuration"
+CONF_EMAILS = "emails"
+CONF_SUBJECTS = "subjects"
+
+CONFIG_SCHEMA = AUTH_PROVIDER_SCHEMA.extend(
+    {
+        vol.Required(CONF_CONFIGURATION): str,
+        vol.Required(CONF_CLIENT_ID): str,
+        vol.Required(CONF_CLIENT_SECRET): str,
+        vol.Optional(CONF_EMAILS, default=[]): [str],
+        vol.Optional(CONF_SUBJECTS, default=[]): [str],
+    },
+    extra=vol.PREVENT_EXTRA,
+)
+
+OPENID_CONFIGURATION_SCHEMA = vol.Schema(
+    {
+        vol.Required("issuer"): str,
+        vol.Required("jwks_uri"): str,
+        vol.Required("id_token_signing_alg_values_supported"): list,
+        vol.Optional("scopes_supported"): vol.Contains("openid"),
+        vol.Required("token_endpoint"): str,
+        vol.Required("authorization_endpoint"): str,
+        vol.Required("response_types_supported"): vol.Contains("code"),
+        vol.Optional(
+            "token_endpoint_auth_methods_supported", default=["client_secret_basic"]
+        ): vol.Contains("client_secret_post"),
+        vol.Optional(
+            "grant_types_supported", default=["authorization_code", "implicit"]
+        ): vol.Contains("authorization_code"),
+    },
+    extra=vol.ALLOW_EXTRA,
```

Comment:
Is this necessary?

Reply:
yes, we use this schema to validate the json we get from configuration endpoint. It can contain much more.